Use one page buffer per column rather than one buffer per page

zolstein opened the pull request:
Replace the columnWriter pages slice of buffers with a single pageBuffer that contains data for all pages in the row-group.

This addresses the fact that when using file buffers, large numbers of files are held open, which can cause errors. It also places more data in contiguous chunks, which should make writing faster.

This commit also adds a new in-memory buffer pool implementation, `chunkMemoryBufferPool`, with memory access characteristics similar to the slice of buffers, and makes it the default pool type for writers. When using a single page buffer per column, this provides significantly better performance than the existing `memoryBufferPool` implementation.

zolstein commented on writer.go:
```diff
@@ -1232,20 +1235,29 @@ func (c *writerColumn) flushFilterPages() error {
-		rbuf.Reset(p)
+	decoder := thrift.NewDecoder(c.header.protocol.NewReader(pageReader))
 
+	for i := 0; i < c.numPages; i++ {
```
This is the only usage of numPages, and it doesn't _seem_ like it should be necessary - instead it seems like we should be able to just iterate through the whole pageBuffer and stop when we finish. However, I'm not sure the best way to construct that loop and handle error cases.

fpetkovski replied:
In theory we should be able to copy raw bytes if pages are already properly constructed into the write buffer.

zolstein replied:
Can you elaborate on this? I don't follow what you're getting at.

fpetkovski replied:
What I was thinking is that the write buffer already has encoded and compressed pages, so what is stopping us from just copying the write buffer as is into the target file/writer?

zolstein replied:
That's what this PR does on line 868. However, here we're populating the bloom filter, which requires decoding the pages so that the right values are added.

fpetkovski replied:
Got it, this makes sense to me 👍

zolstein commented on writer.go:
```diff
@@ -1103,8 +1107,9 @@ func (wb *writerBuffers) swapPageAndScratchBuffers() {
-	pool  BufferPool
-	pages []io.ReadWriteSeeker
+	pool       BufferPool
+	pageBuffer io.ReadWriteSeeker
```
Note: one non-obvious change here is that the pages seemed to hold an invariant that the readers/writers would seek back to zero when finished, but this PR changes this to force consumers to seek to zero before reading.

zolstein commented on writer.go:
```diff
@@ -1232,20 +1235,29 @@ func (c *writerColumn) flushFilterPages() error {
+		return err
+	}
+
+	if _, ok := pageReader.(*os.File); ok {
```
I think this is only useful in the case that pageReader is an os.File? If I'm right, this wrapping should perhaps be baked into the fileBufferPool implementation? It might benefit when reading/writing in other places too.

achille-roussel replied:
I don't know if we can make this assumption as there could be custom implementations of the page buffer which don't use `*os.File` but still benefit from buffering small reads.

If we want to disable buffering we should have an opt-out mechanism rather than opt-in as it's usually the desirable behavior. But unless we have a benchmark showing the benefits I would suggest we avoid early optimizing and keep the code simple.

zolstein replied:
My thinking is that if a custom page buffer implementation benefits from buffering small reads, it should internally use a bufio wrapper.

zolstein reviewed (commented):
Note: this is posted as a draft PR because I haven't had time to really vet this, but hoped to get feedback and make sure it was visible to anyone else who might otherwise have put in the work.

fpetkovski commented on writer.go:
```diff
@@ -859,10 +860,13 @@ func (w *writer) writeRowGroup(rowGroupSchema *Schema, rowGroupSortingColumns []
+		} else if offset != 0 {
+			return 0, fmt.Errorf("resetting parquet page buffer to the start expected offset zero but got %d", offset)
+		}
+		if _, err := io.Copy(&w.writer, c.pageBuffer); err != nil {
```
We could use `io.CopyBuffer` and reuse the buffer across columns instead of using a new one for each column.

zolstein replied:
Not a bad idea, though io.Copy (and CopyBuffer) both use WriterTo or ReaderFrom if the src/dst implement them. I don't know how often this actually winds up using the buffer during the copy anyway, and we'd want to avoid allocating a buffer at all if it won't wind up being used.

fpetkovski replied:
Make sense.

fpetkovski commented on writer.go:
```diff
@@ -1232,20 +1235,29 @@ func (c *writerColumn) flushFilterPages() error {
 
-	rbuf, pool := getBufioReader(nil, 1024)
+	var pageReader io.Reader = c.pageBuffer
+	if _, err = c.pageBuffer.Seek(0, io.SeekStart); err != nil {
```
Do we want to check if the returned `offset` == 0 as in the previous implementation?

zolstein replied:
I can add that back, though I was confused by those checks - if Seek(0, io.SeekStart) didn't set the offset to zero, shouldn't an error have occurred?

fpetkovski replied:
Not sure what the intent was here, maybe @achille-roussel has more context on why the explicit check was in place.

achille-roussel replied:
I think the check was just to ensure that custom implementation of the interface behave as expected, it was a safeguard but shouldn't affect the logic in any other ways.

fpetkovski commented on writer.go:
```diff
@@ -1498,28 +1506,27 @@ func (w *writerColumn) writePageToFilter(page Page) (err error) {
-			c.pool.PutBuffer(buffer)
+func (c *writerColumn) writePageTo(size int64, writeTo func(io.Writer) (int64, error)) (err error) {
+	if c.pageBuffer == nil {
+		c.pageBuffer = c.pool.GetBuffer()
```
I wonder if it makes sense to pass in `size` into `GetBuffer` to hint the allocator how big of a buffer it should create.

zolstein replied:
Actually, if that was an option, it might have changed this design significantly. My assumption was that because this is in the public API we didn't want to make changes that would break code using this library.

Based on the way this buffer is being used here, we'd want to size the buffer for the whole column - I'm not sure that can be known in general. It might be better to just try to use buffer implementations that can grow gracefully.

fpetkovski replied:
This was possible previously because the buffer was holding one page, and the size of the page is known. For the entire column I agree we can't really know the exact size.

fpetkovski reviewed (commented):
Already looks good from my side, but it would be great for someone with deeper knowledge of the library to take a look.

fpetkovski commented:
I updated our project to use this branch and ran the test suite, everything worked out 👍

fpetkovski commented:
I tested this in staging and I noticed this error:

```
panic: runtime error: index out of range [0] with length 0

goroutine 4861030 [running]:
github.com/parquet-go/parquet-go.(*mergeBuffer).read(0xc00ad26000)
	/go/pkg/mod/github.com/zolstein/segio-parquet-go@v0.0.0-20230907175233-758a98d3f16a/merge.go:463 +0x1da
github.com/parquet-go/parquet-go.(*mergeBuffer).WriteRowsTo(0xc00ad26000, {0x25e1320, 0xc004e6e9c0})
	/go/pkg/mod/github.com/zolstein/segio-parquet-go@v0.0.0-20230907175233-758a98d3f16a/merge.go:434 +0x6b
github.com/parquet-go/parquet-go.(*mergedRowGroupRows).WriteRowsTo(0xc004e6ea20, {0x25e1320, 0xc004e6e9c0})
	/go/pkg/mod/github.com/zolstein/segio-parquet-go@v0.0.0-20230907175233-758a98d3f16a/merge.go:114 +0x1aa
github.com/parquet-go/parquet-go.copyRows({0x25e1320, 0xc004e6e9c0}, {0x7f7dbdb3b728, 0xc004e6ea20}, {0x0, 0x0, 0x0})
	/go/pkg/mod/github.com/zolstein/segio-parquet-go@v0.0.0-20230907175233-758a98d3f16a/row.go:319 +0x3f5
github.com/parquet-go/parquet-go.CopyRows(...)
	/go/pkg/mod/github.com/zolstein/segio-parquet-go@v0.0.0-20230907175233-758a98d3f16a/row.go:293
```

fpetkovski commented:
Looking into this further, there was a change in behavior introduced in https://github.com/parquet-go/parquet-go/pull/31 which is causing my application to crash when using merged row groups. I don't think the panic is related to this PR.

achille-roussel commented:
Thanks for looking into it @fpetkovski!

Is the behavior change you referred to a bug that was introduced in parquet-go or caused by assumption that your program was making about the internals? If it's the former, would you mind opening an issue to track it?

achille-roussel reviewed (commented):
A trade off we're making with this change is in the case of using in-memory buffers, the program will need large regions of contiguous memory whereas historically the data would be sliced into pages; this likely will result in higher memory utilization and more memory waste since the column size.

I don't think it's necessarily a reason to avoid making this change, but it would probably be useful to document on the `NewPageBufferPool` function, and maybe a follow up could be to implement a paged memory buffer to use memory more effectively.

fpetkovski commented:
I tried to describe the panic I found in this issue https://github.com/parquet-go/parquet-go/issues/66. I am not sure whether to classify it as a bug, I think we might be able to solve it on our end.

fpetkovski commented:
> A trade off we're making with this change is in the case of using in-memory buffers, the program will need large regions of contiguous memory whereas historically the data would be sliced into pages; this likely will result in higher memory utilization and more memory waste since the column size.

This would indeed be the case with the current implementation. I wonder if it's worth considering a strategy where the memory buffer could allocate and hold multiple byte slices internally instead of growing and copying the existing one when new pages are written. There doesn't have to be a one to one mapping between a byte slice and a page, we could still fit more pages into a single slice.

zolstein commented:
I added a couple of implementations of buffers that internally use a slice-of-byte-slice to allow appending without lots of realloc-and-copies.

My first implementation (chunkMemoryBuffer) used fixed-size buffers. However, it still had a (much smaller, but noticeable) performance regression relative to having one buffer per page as before this whole change. I wound up writing another version (allocMemoryBuffer) that always allocates a new buffer for every write, and the performance seems to be about the same as before. (This version winds up allocating multiple buffers per page, rather than just one, but it avoids doing any reallocs because none of the underlying buffers ever grow.)

I'm really perplexed by why the chunkMemoryBuffer didn't work well, though. It winds up spending much more time in `memmove` (which underlies the `copy` function) despite moving the same amount of data around and generally in chunks of the same size. (Note that the `chunkMemoryBufer.Write` spends significant time on four different lines of `memmove` whereas the others only spend significant time one one line - and they take much less time overall.) My best guess is that it has to do with memory alignment, but I can't say for sure. It's a shame, because the chunkMemoryBuffer seems to have performance advantages in other parts of the code, so if we could fix the `copy` problem it could be a big win. Here are the profiles comparing the old implementation, chunkMemoryBuffers, and allocMemoryBuffers:

<details>
<summary>Separate memory buffer per page</summary>

![profile001](https://github.com/parquet-go/parquet-go/assets/7101542/df7cb8d8-22c1-4876-aac3-640b7a42f904)

</details>

<details>
<summary>chunkMemoryBuffer</summary>

![profile003](https://github.com/parquet-go/parquet-go/assets/7101542/19ccc198-c122-4db7-8bc1-2493b5f9c1e6)

</details>

<details>
<summary>allocMemoryBuffer</summary>

![profile005](https://github.com/parquet-go/parquet-go/assets/7101542/721137e4-11a1-488e-a14c-48f6fd56365c)

</details>

zolstein commented:
I did some more investigation, and it seems like the large performance difference between the two versions is... inconsistent across platforms. I produced it running on a Macos machine, but couldn't reproduce it on a different machine running Linux, or even on the original Macos machine when running under Docker. (And in these situations, it performed better overall due to speed improvements when reading from the buffer.) Given this, I'm not sure if the speed difference while writing is important.

In doing more research, I'm also curious about the prospect of a buffer implementation using an arena allocator (either the experimental version added in 1.20 or a custom allocator) but I haven't figured out a fast custom implementation yet, I don't think we'd want to rely on an experimental feature, and we'd need a fallback for older Go versions anyway.

gernest commented:
> In doing more research, I'm also curious about the prospect of a buffer implementation using an arena allocator (either the experimental version added in 1.20 or a custom allocator)

@zolstein I'm a bit lost here, does `1.20 ` refer to go version ?  The `arena allocator` you mention is it in parquet-go or somewhere else? 

I'm a bit late on this thread, I might have missed something.

zolstein commented:
> @zolstein I'm a bit lost here, does `1.20 ` refer to go version ? The `arena allocator` you mention is it in parquet-go or somewhere else?

Go v1.20 introduced an experimental `arena` package that provides an arena allocator. However, it's an experimental feature, and isn't available at all in earlier versions. Separately, I've tried to see if I could write an arena allocator that would work well here, but I haven't yet managed to write a version that performs well. If I were to do that (now or in the future) my assumption is it would live in `parquet-go`.

gernest commented:
>Go v1.20 introduced an experimental arena package 

@zolstein  Do you have a link to  this package?  I can't find it.

>  Separately, I've tried to see if I could write an arena allocator that would work well here, but I haven't yet managed to write a version that performs well

So , by performance are you referring to raw write throughput? 

I think maybe focusing on smaller gains will be great, If we can reduce the number of files we use for writes, I don't mind the memory penalty we have to pay for it. Then there should be a away to make the larger continuous allocation manageable or even configurable if need be.

zolstein commented:
> @zolstein Do you have a link to this package? I can't find it.

https://go.dev/src/arena/arena.go

Re: performance, my goal was really just that this refactor didn't come with a performance penalty when using in-memory buffers rather than file buffers, especially since that seems like the default use-case. The pre-existing memoryBuffers definitely performed worse, and my first attempt to fix it with the chunkedMemoryBuffer seemed to, though in later testing I've only reproduced it on one computer under MacOS. At this point, it seems like using either alternative buffer implementation, the performance is at least _close_ to the performance without this refactor, so it's just a matter of figuring out what exactly should be merged. There's probably room to optimize this, though, and it might be worth it since this seems like one of a few hot spots in the write path.

thorfour commented:
We heavily rely on in memory buffers for file reading so I am concerned with any significant regressions with a change like this. 

Is it not possible to have this be a file option?

zolstein commented:
@thorfour This change should only affect writing files, not reading them. According to my testing, using one of the two new buffer implementations shouldn't come with a significant regression. If you could profile in your context, though, that could provide valuable data.

The problem with trying to make this a file option is that the only difference between in-memory and file is a different implementation of an interface. Trying to preserve the exact old behavior for in-memory buffers while getting the better behavior for file buffers would require a lot of type-switching, which gets complicated and can only support the buffer implementations that we know about. We could probably solve that by adding an extra layer of abstraction, but we've already exposed a particular API for the BufferPool that makes building the right abstractions difficult, and modifying that could break backward compatibility. I'm also not convinced there's a lot more performance to squeeze out here.

thorfour commented:
> @thorfour This change should only affect writing files, not reading them. According to my testing, using one of the two new buffer implementations shouldn't come with a significant regression. If you could profile in your context, though, that could provide valuable data.
> 
> The problem with trying to make this a file option is that the only difference between in-memory and file is a different implementation of an interface. Trying to preserve the exact old behavior for in-memory buffers while getting the better behavior for file buffers would require a lot of type-switching, which gets complicated and can only support the buffer implementations that we know about. We could probably solve that by adding an extra layer of abstraction, but we've already exposed a particular API for the BufferPool that makes building the right abstractions difficult, and modifying that could break backward compatibility. I'm also not convinced there's a lot more performance to squeeze out here.

Ah apologies for misunderstanding. I will benchmark the changes against our use cases and get back you! Thanks!

zolstein commented:
Sorry for dropping off the face of the earth. Life happened.

I cleaned this up a bit to make it (hopefully) ready to merge. Mainly, I deleted the allocMemoryBuffer/Pool and made the chunkMemBufferPool the default, since I was never able to replicate the performance regression on any machine besides the first. Happy to make any other changes you think are necessary.

achille-roussel commented:
Thanks for following up @zolstein!

@thorfour, do you still want to benchmark this change before we merge it in?

achille-roussel commented:
@thorfour, let us know when we can merge this!

thorfour commented:
@achille-roussel @asubiotto performed a benchmark for us and didn't find any regressions so good to merge on our end! 👍

achille-roussel commented:
Fantastic!

Thanks for all your contributions everyone 🙌
